### PR TITLE
USHIFT-156: csi plugin implementation

### PR DESCRIPTION
**Pull request by copejon**
This PR refactors the default storage provisioner by removing the hostpath provisioner and replacing it with the lvm csi plugin.

**Which issue(s) this PR addresses**:
[USHIFT-156](https://issues.redhat.com//browse/USHIFT-156)

### CODE REVIEW

File: assets/components/odf-lvm/topolvm-openshift-storage_namespace.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: openshift-storage
+  annotations:
+    openshift.io/node-selector: ""
+    workload.openshift.io/allowed: "management"
+  labels:
+    # ODF-LVM should not attempt to manage openshift or kube infra namespaces
+    topolvm.cybozu.com/webhook: "ignore"
```
Comment (dhellmann) [inline assets/components/odf-lvm/topolvm-openshift-storage_namespace.yaml:10]: I don't understand what this is doing. Why do we need so many things to have this label and why shouldn't openshift or kube-infra namespaces be seen by TopoLVM?

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml [hunk 2]
```diff
@@ -0,0 +1,83 @@
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: logicalvolumes.topolvm.cybozu.com
```
Comment (dhellmann) [inline assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: What's the current schedule for this API changing? We should make sure we have a jira ticket to figure out how to handle the transition within MicroShift.

File: pkg/components/storage.go [hunk 3]
```diff
@@ -5,57 +5,111 @@ import (
...
+			"assets/components/odf-lvm/topolvm-controller_deployment.yaml",
 		}
 		sc = []string{
-			"assets/components/hostpath-provisioner/storageclass.yaml",
+			"assets/components/odf-lvm/topolvm_default-storage-class.yaml",
 		}
 	)
+	if err := assets.ApplyStorageClasses(sc, nil, nil, kubeconfigPath); err != nil {
+		klog.Warningf("Failed to apply storage cass %v: %v", sc, err)
+		panic(err)
```
Comment (dhellmann) [inline pkg/components/storage.go:61]: Were these panic changes left over from debugging?

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.name
+        - name: NAMESPACE
+          valueFrom:
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+        image: registry.redhat.io/openshift4/ose-csi-external-provisioner@sha256:42563eb25efb2b6f277944b627bea420fa58fe950b46a1bd1487122b8a387e75
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml:86]: Using a string template can wait for another PR

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        volumeMounts:
+        - mountPath: /run/topolvm
+          name: socket-dir
+        - mountPath: /certs
+          name: certs
+      - args:
+        - --csi-address=/run/topolvm/csi-topolvm.sock
+        - --enable-capacity
+        - --capacity-ownerref-level=2
+        - --capacity-poll-interval=30s
```
[the remark this answers is not in the document]
Comment (copejon) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml:70]: I believe this is related to [CSIStorageCapacity](https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/csi-storage-capacity-v1/) tracking.  This feature allows us to exclude the topolvm extended scheduler.

File: pkg/components/storage.go (hunk 3 above)
Comment (copejon) [reply pkg/components/storage.go]: fixed

**Comment by copejon** [comment]
Holding to allow for reviews

/hold

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            fieldRef:
+              apiVersion: v1
+              fieldPath: metadata.namespace
+        image: registry.redhat.io/openshift4/ose-csi-external-provisioner@sha256:42563eb25efb2b6f277944b627bea420fa58fe950b46a1bd1487122b8a387e75
+        imagePullPolicy: IfNotPresent
+        name: csi-provisioner
+        resources:
+          limits:
+            cpu: 100m
+            memory: 100Mi
```
[the remark this answers is not in the document]
Comment (copejon) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml]: Fixed 👍🏻

File: assets/components/odf-lvm/topolvm-openshift-storage_namespace.yaml (hunk 1 above)
Comment (copejon) [reply assets/components/odf-lvm/topolvm-openshift-storage_namespace.yaml:10]: Actually, I believe this is only a requirement when running the extended scheduler.  Will remove.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml [hunk 7]
```diff
@@ -1,9 +1,13 @@
...
 kind: StorageClass
 metadata:
-  name: kubevirt-hostpath-provisioner
   annotations:
     storageclass.kubernetes.io/is-default-class: "true"
-provisioner: kubevirt.io/hostpath-provisioner
+  labels:
+  name: topolvm-provisioner
+parameters:
+  csi.storage.k8s.io/fstype: xfs
```
Comment (dhellmann) [inline assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: Is this something users will eventually want to configure here or would we suggest that they create a separate StorageClass to do that?

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml [hunk 8]
```diff
@@ -1,9 +1,13 @@
 apiVersion: storage.k8s.io/v1
 kind: StorageClass
 metadata:
-  name: kubevirt-hostpath-provisioner
   annotations:
     storageclass.kubernetes.io/is-default-class: "true"
-provisioner: kubevirt.io/hostpath-provisioner
+  labels:
+  name: topolvm-provisioner
```
Comment (dhellmann) [inline assets/components/odf-lvm/topolvm_default-storage-class.yaml:7]: Is this name significant to the code? Should it be a variation of ODF-LVM to be consistent with branding elsewhere?

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml (hunk 2 above)
Comment (copejon) [reply assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: The API's groupVersionKind is expected to change in or by Q4 this year to replace the maintaining company's name.  There's not a set date for this yet.  General versioning has a spike already, but it'll be worth it to have this upcoming change tracked for awareness I think.

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (dhellmann) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml:70]: Does it need to run so often? Is it something users are likely going to need to be able to configure to manage the performance of the system?

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml (hunk 2 above)
Comment (dhellmann) [reply assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: I'll be very interested to hear what the ODF team's plan is for rolling upgrades with a GVK name change like that.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml (hunk 7 above)
Comment (copejon) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: The storageClass is typically a cluster-admin defined API, so it's reasonable to assume to microshift users would want to specify this.

Comment (dhellmann) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: OK, so we don't need to provide a config option for this, but we might need to give documentation for how to create an alternative storage class.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml (hunk 8 above)
Comment (copejon) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:7]: No, the storageClass name can be arbitrary, as can the deployment and daemonset, and other included APIs.  

WRT branding, on OCP the ODF LVM Operator deploys all the operand APIs (pods, configMaps, secrets, role, roleBindings, etc) with a `topolvm-` prefix.  Functionally, this is a trivial string replacement in rebase automation.

I'm a little concerned about renaming a OCP component for the sake of branding, which could create the appearance of different LVM provisioning solutions between OCP and MicroShift.  I'll defer to your take on that though.

Comment (dhellmann) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:7]: If the LVMO uses the `topolvm-` prefix elsewhere, I agree we should stay consistent with that. Thanks for explaining the reasoning here. We should try to mention this decision in the enhancement as well.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml (hunk 7 above)
Comment (copejon) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: +1

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml (hunk 2 above)
Comment (copejon) [reply assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: I'm following up with them internally.  Upstream, there's a plan to implement a migration tool, but I don't know what plans for downstream users will be.

File: assets/components/odf-lvm/topolvm-node_daemonset.yaml [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+      labels:
+        app: topolvm-node
+      name: lvmcluster-sample
+    spec:
+      containers:
+      - command:
+        - /lvmd
+        - --config=/etc/topolvm/lvmd.yaml
+        - --container=true
+        image: registry.redhat.io/odf4/odf-topolvm-rhel8@sha256:bd9fb330fc35f88fae65f1598b802923c8a9716eeec8432bdf05d16bd4eced64
```
Comment (mangelajo) [inline assets/components/odf-lvm/topolvm-node_daemonset.yaml:23]: image reference templates here too.

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml (hunk 8 above)
Comment (mangelajo) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:7]: Maintaining consistency makes sense to me too

File: assets/components/odf-lvm/topolvm_default-storage-class.yaml (hunk 7 above)
Comment (mangelajo) [reply assets/components/odf-lvm/topolvm_default-storage-class.yaml:9]: :+1

File: assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml (hunk 2 above)
Comment (mangelajo) [reply assets/crd/0000_20_topolvm.cybozu.com_logicalvolumes.yaml:4]: May be a web hook in the meanwhile to provide translation between both versions?

File: pkg/release/release_arm64.go [hunk 10]
```diff
@@ -21,13 +21,12 @@ package release
...
-		"pause":                         "k8s.gcr.io/pause:3.6",
-		"service_ca_operator":           "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:97a7d56ac6d6a38f7977b0759bb1b97fe3895e6020364e60a8a2926f3fd0084e",
-		"ovn_kubernetes":                "quay.io/microshift/ovn-kubernetes-singlenode:4.10.18",
+		"cli":                 "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:65889fde898f3025462f1344bd93e4a1d8ae5a764224b61b4f0b7e46cc1108e4",
+		"coredns":             "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:2619cbcbc5ea990bd8349848d410cc4040aea0265821b9d1e46b364f0519481b",
+		"haproxy_router":      "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:972f3fadc1b7eab2eaa02eff4128149eabe9577943dcb3e56f44dad7fc14059e",
+		"kube_rbac_proxy":     "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:2d8319a33aa781918a3173e2061635d8534cc5e4a39171775830ff7e2ce646ac",
+		"pause":               "k8s.gcr.io/pause:3.6",
+		"service_ca_operator": "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:97a7d56ac6d6a38f7977b0759bb1b97fe3895e6020364e60a8a2926f3fd0084e",
```
Comment (mangelajo) [inline pkg/release/release_arm64.go:29]: If you ask @fzdarsky he can share how to find the arm64 images.

**Review by mangelajo — CHANGES_REQUESTED** [review]
some tiny comments, other than that it looks good

File: assets/components/odf-lvm/topolvm-controller_deployment.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhellmann) [reply assets/components/odf-lvm/topolvm-controller_deployment.yaml:83]: I'd be OK doing that in a follow-up PR if that helps us keep moving.

**Comment by dhellmann** [comment]
/lgtm

Let's take this and deal with the cleanup items as follow-ups.

**Pull request merged**